### PR TITLE
Add debug (non-whitespace) version of calc grammar 

### LINKED ISSUE #35: Different versions of the grammars
It would be useful (for demonstrations and debugging) to have a version of the `calc` grammar with no whitespace, because the full trees are difficult to read.

### PR DESCRIPTION
**Pull request by snim2**
Intended to simplify demonstration and debugging.

Fixes #35 

With new `--grammar` switch:

  * [map.pdf](https://github.com/softdevteam/diffract/files/1352868/map.pdf)
  * [edit.pdf](https://github.com/softdevteam/diffract/files/1352870/edit.pdf)

Without `--grammar`:

  * [map-full.pdf](https://github.com/softdevteam/diffract/files/1352873/map-full.pdf)
  * [edit-full.pdf](https://github.com/softdevteam/diffract/files/1352875/edit-full.pdf)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -69,6 +69,8 @@ Options:
     -d, --dot <file>    write out GraphViz representations of the input file(s).
     --edit <file>       write a GraphViz representation of the mapping store to
                         file, after the edit script has been generated.
+    --grammar           use the 'debug' version of the relevant grammar, if
```
Comment (ltratt) [inline src/main.rs]: This seems like a weird name, and weird thing, for the option to do. Why would I want a debug grammar? I suspect I'm missing out on something!

File: src/main.rs [hunk 2]
```diff
@@ -283,21 +286,60 @@ fn main() {
...
+        canonicalize(dir.join(format!("../../grammars/debug-{}.l", ext1)))
+            .unwrap_or_else(|_| canonicalize(dir.join("../../grammars/txt.l")).unwrap())
+    } else {
+        canonicalize(dir.join(format!("../../grammars/{}.l", ext1)))
+            .unwrap_or_else(|_| canonicalize(dir.join("../../grammars/txt.l")).unwrap())
+    };
+    let yacc_y_path1 = if args.flag_grammar &&
+                          Path::new(&canonicalize(dir.join(format!("../../grammars/debug-{}.y",
+                                                                   ext1)))
+                                        .unwrap())
```
Comment (ltratt) [inline src/main.rs]: The indentation means I'm not sure what this is unwrapping.

File: src/main.rs [hunk 3]
```diff
@@ -283,21 +286,60 @@ fn main() {
...
         }
     };
-    let lex_l_path2 = canonicalize(dir.join(format!("../../grammars/{}.l", ext2)))
-                        .unwrap_or_else(|_| canonicalize(dir.join("../../grammars/txt.l")).unwrap());
-    let yacc_y_path2 = canonicalize(dir.join(format!("../../grammars/{}.y", ext2)))
-                        .unwrap_or_else(|_| canonicalize(dir.join("../../grammars/txt.y")).unwrap());
+    let lex_l_path2 = if args.flag_grammar &&
+                         Path::new(&canonicalize(dir.join(format!("../../grammars/debug-{}.l",
+                                                                  ext2)))
+                                       .unwrap())
```
Comment (ltratt) [inline src/main.rs]: Same comment as above.

**Comment by snim2** [comment]
Ready for squashing?

**Comment by ltratt** [comment]
Obvious question: what's the difference between having an extra flag and the user just passing in a different grammar normally?

**Comment by snim2** [comment]
There is no option to pass in a grammar with Diffract. Instead, Diffract looks at the file extension of your input files, and if it has a grammar that matches it uses that. If not it falls back to the "plain text" grammar.

**Comment by ltratt** [comment]
OK, so perhaps we should think of this as a "grammar override" option (not a debug grammar). Even if diffract has a gramar for file X, you can override it with this; but you could also use it to have diffract parse languages it doesn't natively know about.

**Comment by snim2** [comment]
Well ... that is not how this PR works currently. I've been careful not to assume that the user wishes to diff two files in the same language, so if we want to have a "use my grammar" / "use my lexer" option, then we would potentially have either two or four different switches for lexers / parsers for the two input files.

**Comment by ltratt** [comment]
I understand the PR doesn't have this functionality yet, but I think we need to think about it because it's much more general.

How about if we had an option called (say) `-g` which takes one or two pairs. e.g. `-g java.l,java.y` forces both input files to be parsed as Java but `-g java.l,java.y -g php.l,php.y` forces the "before" file to be parsed as Java and the "after" file as PHP. I'm not saying this is the best possible interface, but I'm trying to think of how we can generalise this in a nice way.

**Comment by snim2** [comment]
I don't especially object to that for debugging functionality, but I think an ordinary end-user should never have to know about parsers and lexers. Do you want this change to go in this PR?

**Comment by ltratt** [comment]
I'm fine with the "choose a grammar by default" which Diffract currently has (indeed, I think it's great!). The suggestion I'm making is for advanced users who have a good reason to use a different lexer/parser.

**Comment by snim2** [comment]
OK, well we might as well use this PR, I'll have a go!

**Comment by ltratt** [comment]
Thanks!

**Comment by snim2** [comment]
OK, this works, modulo the bug I just raised in #39

**Comment by ltratt** [comment]
OK, I think we need to do a bit of squashing then I can merge.

**Comment by snim2** [comment]
Squashed!

**Pull request merged**